File: app/src/main/java/remix/myplayer/lyric/SearchLRC.java
package remix.myplayer.lyric;

import android.database.Cursor;
import android.provider.MediaStore;
import android.text.TextUtils;
import android.util.Base64;

import com.google.gson.Gson;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import remix.myplayer.APlayerApplication;
import remix.myplayer.misc.cache.DiskCache;
import remix.myplayer.misc.cache.DiskLruCache;
import remix.myplayer.model.LrcRequest;
import remix.myplayer.model.mp3.Song;
import remix.myplayer.model.netease.NSongSearchResponse;
import remix.myplayer.util.CommonUtil;
import remix.myplayer.util.Global;
import remix.myplayer.util.LogUtil;
import remix.myplayer.util.SPUtil;

/**
 * Created by Remix on 2015/12/7.
 */

/**
 * 根据歌曲名和歌手名 搜索歌词并解析成固定格式
 */
public class SearchLRC {
    private static final String TAG = "SearchLRC";
    private static final String DEFAULT_LOCAL = "GB2312";
    private static final String LRC_REQUEST_ROOT = "http://s.geci.me/lrc/";
    private ILrcParser mLrcBuilder;
    private Song mInfo;
    private String mTitle;
    private String mArtistName;
    private String mDisplayName;
    private String mManualPath;

    public SearchLRC(Song item) {
        mInfo = item;
        mTitle = mInfo.getTitle();
        mArtistName = mInfo.getArtist();
        try {
            if(!TextUtils.isEmpty(mInfo.getDisplayname())){
                String temp = mInfo.getDisplayname();
                mDisplayName = temp.indexOf('.') > 0 ? temp.substring(0,temp.lastIndexOf('.')) : temp;
            }
        } catch (Exception e){
            CommonUtil.uploadException("SearchLrc Init Error","DisPlayName:" + item.getDisplayname() + " Title:" + item.getTitle());
            mDisplayName = mTitle;
        }
        mLrcBuilder = new DefaultLrcParser();
    }

    public SearchLRC(Song item, String manualPath){
        this(item);
        mManualPath = manualPath;
    }

    public int getSongID(){
        return mInfo.getId();
    }


    /**
     * 获取酷狗歌词接口的参数
     * @return
     */
    public LrcRequest getLrcParam(){
        //酷狗
        try {
            JSONObject response = CommonUtil.getSongJsonObject(URLEncoder.encode(mInfo.getTitle(), "utf-8"),
                            URLEncoder.encode(mInfo.getArtist(), "utf-8"),mInfo.getDuration());
            if(response != null && response.length() > 0){
                if(response.getJSONArray("candidates").length() > 0){
                    JSONObject jsonObject = response.getJSONArray("candidates").getJSONObject(0);
                    if(jsonObject.getInt("score") >= 60)
                        return new LrcRequest(jsonObject.getInt("id"),jsonObject.getString("accesskey"));
                }
            }
        }catch (Exception e){
            e.printStackTrace();
        }
        return new LrcRequest();
    }

    public void test(){
        new Thread(){
            @Override
            public void run() {
                try {


                    Map<String,String> params = new HashMap<>();
                    params.put("s","Seemann");
                    params.put("offset","0");
                    params.put("limit","0");
                    params.put("type","1");
                    StringBuilder paramStr = new StringBuilder();

                    int pos = 0;
                    for(String key : params.keySet()){
                        if(pos > 0)
                            paramStr.append("&");
                        paramStr.append(String.format("%s=%s", key, URLEncoder.encode(params.get(key), "utf-8")));
                        pos++;
                    }

                    final String url = "http://music.163.com/api/search/pc?s=Seemann&offset=0&limit=10&type=1006";
                    OkHttpClient client = new OkHttpClient.Builder().build();
                    RequestBody requestBody = RequestBody.create(MediaType.parse("application/x-www-form-urlencoded; charset=utf-8"),"");

                    Request request = new Request.Builder()
                            .url(url)
                            .post(requestBody)
                            .addHeader("Cookie","appver=1.5.0.75771")
                            .build();

                    client.newCall(request).enqueue(new Callback() {
                        @Override
                        public void onFailure(Call call, IOException e) {

                        }

                        @Override
                        public void onResponse(Call call, Response response) throws IOException {
                            String responseString = response.body().string();
                            NSongSearchResponse searchResponse = new Gson().fromJson(responseString,NSongSearchResponse.class);
                            LogUtil.d("NSongSearchResponse", searchResponse + "");
                        }
                    });

//                    Response response = client.newCall(request).execute();


                } catch (Exception e) {
                    e.printStackTrace();
                }

            }
        }.start();

    }

    /**
     * 根据歌词id,发送请求并解析歌词
     * @return 歌词信息list
     */
    public List<LrcRow> getLrc(String lrcPath){
        //判断是否是忽略的歌词
        Set<String> ignoreLrcId = SPUtil.getStringSet(APlayerApplication.getContext(),"Setting","IgnoreLrcID");
        if(ignoreLrcId != null && ignoreLrcId.size() > 0){
            for (String id : ignoreLrcId){
                if((mInfo.getId() + "").equals(id)){
                    return null;
                }
            }
        }
        BufferedReader br = null;
        //manualPath不为空说明为手动设置歌词
        try {
            if(!TextUtils.isEmpty(lrcPath)){
                br = new BufferedReader(new InputStreamReader(new FileInputStream(lrcPath)));
                return mLrcBuilder.getLrcRows(br,true, mTitle,mArtistName);
            }
        } catch (Exception e){
            LogUtil.e(TAG,e.toString());
        } finally {
            if (br != null)
                try {
                    br.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
        }

        //搜索时判断该歌曲是否有缓存
        try {
            DiskLruCache.Snapshot snapShot = DiskCache.getLrcDiskCache().get(CommonUtil.hashKeyForDisk(mTitle + "/" + mArtistName));
             if(snapShot != null && (br = new BufferedReader(new InputStreamReader(snapShot.getInputStream(0)))) != null ){
                return mLrcBuilder.getLrcRows(br,false, mTitle,mArtistName);
            }
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            if(br != null)
                try {
                    br.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
        }

        //搜索歌词
        try {
            //是否优先搜索在线歌词
            boolean onlineFirst = SPUtil.getValue(APlayerApplication.getContext(),"Setting","OnlineLrc",false);
            if(onlineFirst){
                String onlineLrcContent = getOnlineLrcContent();
                if(!TextUtils.isEmpty(onlineLrcContent)){
                    br = new BufferedReader(
                            new InputStreamReader(new ByteArrayInputStream(Base64.decode(onlineLrcContent, Base64.DEFAULT))));
                    return mLrcBuilder.getLrcRows(br,true, mTitle,mArtistName);
                } else {
                    String localLrcPath = getlocalLrcPath();
                    if(!localLrcPath.equals("")){
                        br = new BufferedReader(new InputStreamReader(new FileInputStream(localLrcPath)));
                        return mLrcBuilder.getLrcRows(br,true, mTitle,mArtistName);
                    }
                }
            } else {
                String localLrcPath = getlocalLrcPath();
                if(!localLrcPath.equals("")){
                    br = new BufferedReader(new InputStreamReader(new FileInputStream(localLrcPath)));
                    return mLrcBuilder.getLrcRows(br,true, mTitle,mArtistName);
                } else {
                    String onlineLrcContent = getOnlineLrcContent();
                    if(!TextUtils.isEmpty(onlineLrcContent)){
                        br = new BufferedReader(
                                new InputStreamReader(new ByteArrayInputStream(Base64.decode(onlineLrcContent, Base64.DEFAULT))));
                        return mLrcBuilder.getLrcRows(br,true, mTitle,mArtistName);
                    }
                }
            }
        }catch (Exception e){
//            LogUtil.e(TAG,e.toString());
            CommonUtil.uploadException("Search Error",e);
        } finally {
            if (br != null)
                try {
                    br.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
        }
        return null;
    }

    /**
     * 获得在线歌词
     * @return
     */
    private String getOnlineLrcContent(){
        LrcRequest lrcParam = getLrcParam();
        BufferedReader br = null;
        if(lrcParam != null && !TextUtils.isEmpty(lrcParam.AccessKey)){
            try {
                URL url = new URL("http://lyrics.kugou.com/download?ver=1&client=pc&id=" + lrcParam.ID + "&accesskey=" + lrcParam.AccessKey + "&fmt=lrc&charset=utf8");
                HttpURLConnection httpConn = (HttpURLConnection) url.openConnection();
                httpConn.setConnectTimeout(10000);
                httpConn.connect();
                br = new BufferedReader(new InputStreamReader(httpConn.getInputStream()));
                StringBuffer stringBuffer = new StringBuffer(128);
                String s;
                while ((s = br.readLine()) != null){
                    stringBuffer.append(s);
                }
                if(TextUtils.isEmpty(stringBuffer)){
                    return null;
                }
                return new JSONObject(stringBuffer.toString()).getString("content");
            } catch (JSONException e) {
                e.printStackTrace();
            } catch (IOException e) {
                e.printStackTrace();
            }  finally {
                if(br != null)
                    try {
                        br.close();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
            }
        }
        return null;
    }

    /**
     * 搜索本地所有歌词文件
     * @return
     */
    private String getlocalLrcPath() {
        //查找本地目录
        String searchPath =  SPUtil.getValue(APlayerApplication.getContext(),"Setting","LrcSearchPath","");
        if(mInfo == null)
            return "";
        if(!TextUtils.isEmpty(searchPath)){
            //已设置歌词路径
            CommonUtil.searchFile(mDisplayName,mTitle,mArtistName, new File(searchPath));
            if(!TextUtils.isEmpty(Global.CurrentLrcPath)){
                return Global.CurrentLrcPath;
            }

        } else{
            //没有设置歌词路径 搜索所有歌词文件
            Cursor allLrcFiles = null;
            try {
                allLrcFiles = APlayerApplication.getContext().getContentResolver().
                        query(MediaStore.Files.getContentUri("external"),
                                null,
                                MediaStore.Files.FileColumns.DATA + " like ? or " +
                                MediaStore.Files.FileColumns.DATA + " like ? or " +
                                MediaStore.Files.FileColumns.DATA + " like ? ",
                                new String[]{"%lyric%","%Lyric%","%.lrc"},
                                null);
                if(allLrcFiles == null || !(allLrcFiles.getCount() > 0))
                    return "";
                while (allLrcFiles.moveToNext()){
                    File file = new File(allLrcFiles.getString(allLrcFiles.getColumnIndex(MediaStore.Files.FileColumns.DATA)));
                    if (file.exists() && file.canRead()) {
                        if(CommonUtil.isRightLrc(file, mDisplayName,mTitle,mArtistName)){
                            return file.getAbsolutePath();
                        }
                    }
                }
            }catch (Exception e){
                e.printStackTrace();
            } finally {
                if(allLrcFiles != null && !allLrcFiles.isClosed())
                    allLrcFiles.close();
            }
        }

        return "";
    }

}

File: app/src/main/java/remix/myplayer/lyric/SearchLrc.kt
package remix.myplayer.lyric

import android.database.Cursor
import android.provider.MediaStore
import android.text.TextUtils
import android.util.Base64
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import io.reactivex.Observable
import io.reactivex.functions.Function
import remix.myplayer.App
import remix.myplayer.bean.kugou.KLrcResponse
import remix.myplayer.bean.kugou.KSearchResponse
import remix.myplayer.bean.misc.LyricPriority
import remix.myplayer.bean.mp3.Song
import remix.myplayer.bean.netease.NLrcResponse
import remix.myplayer.bean.netease.NSongSearchResponse
import remix.myplayer.lyric.bean.LrcRow
import remix.myplayer.misc.cache.DiskCache
import remix.myplayer.misc.tageditor.TagEditor
import remix.myplayer.request.network.HttpClient
import remix.myplayer.request.network.RxUtil
import remix.myplayer.util.*
import java.io.*
import java.nio.charset.Charset
import java.util.*

/**
 * Created by Remix on 2015/12/7.
 */

/**
 * 根据歌曲名和歌手名 搜索歌词并解析成固定格式
 */
class SearchLrc(private val mSong: Song) {
    private val mLrcParser: ILrcParser
    private var mDisplayName: String? = null
    private var mCacheKey: String? = null
    private var mSearchKey: String? = null

    init {
        try {
            if (!TextUtils.isEmpty(mSong.displayname)) {
                val temp = mSong.displayname
                mDisplayName = if (temp.indexOf('.') > 0) temp.substring(0, temp.lastIndexOf('.')) else temp
            }
            mSearchKey = getLyricSearchKey(mSong)
        } catch (e: Exception) {
            LogUtil.d(TAG, e.toString())
            mDisplayName = mSong.title
        }

        mLrcParser = DefaultLrcParser()
    }

    /**
     * 根据歌词id,发送请求并解析歌词
     *
     * @return 歌词
     */
    fun getLyric(manualPath: String, clearCache: Boolean): Observable<List<LrcRow>> {
        val type = SPUtil.getValue(App.getContext(), SPUtil.LYRIC_KEY.NAME, mSong.id, SPUtil.LYRIC_KEY.LYRIC_DEFAULT)

        val observable = when (type) {
            SPUtil.LYRIC_KEY.LYRIC_IGNORE -> {
                LogUtil.d(TAG, "Ignore Lyric")
                Observable.error(Throwable("Ignore Lyric"))
            }
            SPUtil.LYRIC_KEY.LYRIC_EMBEDDED -> {
                getEmbeddedObservable()
            }
            SPUtil.LYRIC_KEY.LYRIC_LOCAL -> {
                getLocalObservable()
            }
            SPUtil.LYRIC_KEY.LYRIC_NETEASE -> {
                getNeteaseObservable()
            }
            SPUtil.LYRIC_KEY.LYRIC_KUGOU -> {
                getKuGouObservable()
            }
            SPUtil.LYRIC_KEY.LYRIC_MANUAL -> {
                getManualObservable(manualPath)
            }
            SPUtil.LYRIC_KEY.LYRIC_DEFAULT -> {
                //默认优先级排序 网易-酷狗-本地-内嵌
                val priority = Gson().fromJson<List<LyricPriority>>(SPUtil.getValue(App.getContext(), SPUtil.LYRIC_KEY.NAME, SPUtil.LYRIC_KEY.PRIORITY_LYRIC, SPUtil.LYRIC_KEY.DEFAULT_PRIORITY),
                        object : TypeToken<List<LyricPriority>>() {}.type)

                val observables = mutableListOf<Observable<List<LrcRow>>>()
                priority.forEach {
                    when (it.priority) {
                        LyricPriority.NETEASE.priority -> observables.add(getNeteaseObservable())
                        LyricPriority.KUGOU.priority -> observables.add(getKuGouObservable())
                        LyricPriority.LOCAL.priority -> observables.add(getLocalObservable())
                        LyricPriority.EMBEDED.priority -> observables.add(getEmbeddedObservable())
                    }
                }
                Observable.concat(observables).firstOrError().toObservable()
            }
            else -> {
                Observable.error<List<LrcRow>>(Throwable("Not support type"))
            }
        }

        return if (isTypeAvailable(type)) Observable.concat(getCacheObservable(), observable)
                .firstOrError()
                .toObservable()
                .doOnSubscribe { disposable ->
                    LOCAL_LYRIC_PATH = ""
                    mCacheKey = Util.hashKeyForDisk(mSong.id.toString() + "-" +
                            (if (!TextUtils.isEmpty(mSong.artist)) mSong.artist else "") + "-" +
                            if (!TextUtils.isEmpty(mSong.title)) mSong.title else "")
                    LogUtil.d(TAG, "CacheKey: $mCacheKey SearchKey: $mSearchKey")
                    if (clearCache) {
                        LogUtil.d(TAG, "clearCache")
                        DiskCache.getLrcDiskCache().remove(mCacheKey)
                    }
                }.compose(RxUtil.applyScheduler())
        else
            observable
    }

    private fun isTypeAvailable(type: Int): Boolean {
        return type != SPUtil.LYRIC_KEY.LYRIC_IGNORE
    }

    /**
     * 根据歌词id,发送请求并解析歌词
     *
     * @return 歌词
     */
    fun getLyric(): Observable<List<LrcRow>> {
        return getLyric("", false)
    }

    /**
     * 内嵌歌词
     *
     * @return
     */
    private fun getEmbeddedObservable(): Observable<List<LrcRow>> {
        return Observable.create { e ->
            val lyric = TagEditor(mSong.url).lyric
            if (!lyric.isNullOrEmpty()) {
                e.onNext(mLrcParser.getLrcRows(getBufferReader(lyric!!.toByteArray(UTF_8)), true, mCacheKey, mSearchKey))
                LogUtil.d(TAG, "EmbeddedLyric")
            }
            e.onComplete()
        }
    }

    /**
     * 缓存
     */
    private fun getCacheObservable(): Observable<List<LrcRow>> {
        return Observable.create { e ->
            DiskCache.getLrcDiskCache().get(mCacheKey)?.run {
                BufferedReader(InputStreamReader(getInputStream(0))).use {
                    it.readLine().run {
                        e.onNext(Gson().fromJson(this, object : TypeToken<List<LrcRow>>() {}.type))
                        LogUtil.d(TAG, "CacheLyric")
                    }
                }
            }
            e.onComplete()
        }
    }


    private val isCN: Boolean
        get() = "zh".equals(Locale.getDefault().language, ignoreCase = true)

    /**
     * 搜索本地所有歌词文件
     * 查找本地目录
     * 已设置歌词路径
     * 没有设置歌词路径 搜索所有歌词文件
     *
     * @return
     */
    private val localLrcPath: String
        get() {
            val searchPath = SPUtil.getValue(App.getContext(), SPUtil.SETTING_KEY.NAME, SPUtil.SETTING_KEY.LOCAL_LYRIC_SEARCH_DIR, "")
            if (!TextUtils.isEmpty(searchPath)) {
                LyricUtil.searchFile(mDisplayName, mSong.title, mSong.artist, File(searchPath))
                if (!TextUtils.isEmpty(SearchLrc.LOCAL_LYRIC_PATH)) {
                    return SearchLrc.LOCAL_LYRIC_PATH
                }
            } else {
                var filesCursor: Cursor? = null
                try {
                    filesCursor = App.getContext().contentResolver.query(MediaStore.Files.getContentUri("external"), null,
                            MediaStore.Files.FileColumns.DATA + " like ? or " +
                                    MediaStore.Files.FileColumns.DATA + " like ? or " +
                                    MediaStore.Files.FileColumns.DATA + " like ? ",
                            arrayOf("%lyric%", "%Lyric%", "%.lrc%"), null)
                    if (filesCursor == null || filesCursor.count <= 0)
                        return ""
                    while (filesCursor.moveToNext()) {
                        val file = File(filesCursor.getString(filesCursor.getColumnIndex(MediaStore.Files.FileColumns.DATA)))
                        if (file.exists() && file.isFile && file.canRead()) {
                            if (LyricUtil.isRightLrc(file, mDisplayName, mSong.title, mSong.artist)) {
                                return file.absolutePath
                            }
                        }
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                } finally {
                    if (filesCursor != null && !filesCursor.isClosed)
                        filesCursor.close()
                }
            }
            return ""
        }

    /**
     * 搜索本地所有歌词文件
     */
    private val allLocalLrcPath: List<String>
        get() {
            val results = ArrayList<String>()
            val searchPath = SPUtil.getValue(App.getContext(), SPUtil.SETTING_KEY.NAME, SPUtil.SETTING_KEY.LOCAL_LYRIC_SEARCH_DIR, "")
            if (!TextUtils.isEmpty(searchPath)) {//已设置歌词搜索路径
                LyricUtil.searchFile(mDisplayName, mSong.title, mSong.artist, File(searchPath))
                if (!TextUtils.isEmpty(LOCAL_LYRIC_PATH))
                    results.add(SearchLrc.LOCAL_LYRIC_PATH)
                return results
            } else {//没有设置歌词路径 搜索所有歌词文件
                App.getContext().contentResolver.query(MediaStore.Files.getContentUri("external"), null,
                        MediaStore.Files.FileColumns.DATA + " like ? or " +
                                MediaStore.Files.FileColumns.DATA + " like ? or " +
                                MediaStore.Files.FileColumns.DATA + " like ? ",
                        arrayOf("%lyric%", "%Lyric%", "%.lrc"), null)
                        .use { filesCursor ->
                            while (filesCursor.moveToNext()) {
                                val file = File(filesCursor.getString(filesCursor.getColumnIndex(MediaStore.Files.FileColumns.DATA)))
                                if (file.exists() && file.isFile && file.canRead()) {
                                    if (LyricUtil.isRightLrc(file, mDisplayName, mSong.title, mSong.artist)) {
                                        results.add(file.absolutePath)
                                    }
                                }
                            }
                            return results
                        }
            }

        }

    /**
     * 网络或者本地歌词
     *
     * @param type
     * @return
     */
    @Deprecated("")
    private fun getContentObservable(type: Int): Observable<List<LrcRow>> {
        val networkObservable = getNetworkObservable(type)
        val localObservable = getLocalObservable()
        val onlineFirst = SPUtil.getValue(App.getContext(), SPUtil.SETTING_KEY.NAME, SPUtil.SETTING_KEY.ONLINE_LYRIC_FIRST, false)
        return Observable.concat(if (onlineFirst) networkObservable else localObservable, if (onlineFirst) localObservable else networkObservable).firstOrError().toObservable()
    }

    /**
     * 手动设置歌词
     */
    private fun getManualObservable(manualPath: String): Observable<List<LrcRow>> {
        return Observable.create { e ->
            //手动设置的歌词
            if (!TextUtils.isEmpty(manualPath)) {
                LogUtil.d(TAG, "ManualLyric")
                e.onNext(mLrcParser.getLrcRows(getBufferReader(manualPath), true, mCacheKey, mSearchKey))
            }
            e.onComplete()
        }
    }


    /**
     * 本地歌词
     *
     * @param type
     * @return
     */
    private fun getLocalObservable(): Observable<List<LrcRow>> {
        return Observable.create { e ->
            val localPaths = ArrayList<String>(allLocalLrcPath)
            if (localPaths.isNotEmpty()) {
                if (localPaths.size > 1 && isCN) {
                    localPaths.sortWith(Comparator { o1, o2 -> o2.compareTo(o1) })
                    val localPath = localPaths[0]
                    var translatePath: String? = null
                    for (path in localPaths) {
                        if (path.contains("translate") && path != localPath) {
                            translatePath = path
                            break
                        }
                    }
                    LogUtil.d(TAG, "LocalLyric")
                    if (translatePath == null) {
                        e.onNext(mLrcParser.getLrcRows(getBufferReader(localPath), true, mCacheKey, mSearchKey))
                    } else {
                        //合并歌词
                        val source = mLrcParser.getLrcRows(getBufferReader(localPath), false, mCacheKey, mSearchKey)
                        val translate = mLrcParser.getLrcRows(getBufferReader(translatePath), false, mCacheKey, mSearchKey)
                        if (translate != null && translate.size > 0) {
                            //                                    int j = 0;
                            //                                    for (int i = 0; i < source.size(); ) {
                            //                                        boolean match = Math.abs(translate.get(j).getTime() - source.get(i).getTime()) < 1000;
                            //                                        if (match) {
                            //                                            source.get(i).setTranslate(translate.get(j).getContent());
                            //                                            i++;
                            //                                        } else if(translate.get(j).getTime() > source.get(i).getTime()){
                            //                                            i++;
                            //                                        } else {
                            //                                            j++;
                            //                                        }
                            //                                    }
                            for (i in translate.indices) {
                                for (j in source.indices) {
                                    if (isTranslateCanUse(translate[i].content) && translate[i].time == source[j].time) {
                                        source[j].translate = translate[i].content
                                        break
                                    }
                                }
                            }
                            mLrcParser.saveLrcRows(source, mCacheKey, mSearchKey)
                            e.onNext(source)
                        }
                    }
                } else {
                    val localPath = localPaths[0]
                    e.onNext(mLrcParser.getLrcRows(getBufferReader(localPath), true, mCacheKey, mSearchKey))
                }
            }
            e.onComplete()
        }
    }

    /**
     * 网易歌词
     */
    private fun getNeteaseObservable(): Observable<List<LrcRow>> {
        return HttpClient.getNeteaseApiservice()
                .getNeteaseSearch(mSearchKey, 0, 1, 1)
                .flatMap { it ->
                    HttpClient.getInstance()
                            .getNeteaseLyric(Gson().fromJson(it.string(), NSongSearchResponse::class.java).result.songs.get(0).id)
                            .map {
                                val lrcResponse = Gson().fromJson(it.string(), NLrcResponse::class.java)
                                val combine = mLrcParser.getLrcRows(getBufferReader(lrcResponse.lrc.lyric.toByteArray()), false, mCacheKey, mSearchKey)
                                if (isCN && lrcResponse.tlyric != null && !TextUtils.isEmpty(lrcResponse.tlyric.lyric)) {
                                    val translate = mLrcParser.getLrcRows(getBufferReader(lrcResponse.tlyric.lyric.toByteArray()), false, mCacheKey, mSearchKey)
                                    if (translate != null && translate.size > 0) {
                                        for (i in translate.indices) {
                                            for (j in combine.indices) {
                                                if (translate[i].time == combine[j].time) {
                                                    combine[j].translate = translate[i].content
                                                    break
                                                }
                                            }
                                        }
                                    }
                                }
                                LogUtil.d(TAG, "NeteaseLyric")
                                mLrcParser.saveLrcRows(combine, mCacheKey, mSearchKey)
                                combine
                            }
                }.onErrorResumeNext(Function {
                    Observable.empty()
                })
    }

    /**
     * 酷狗歌词
     */
    private fun getKuGouObservable(): Observable<List<LrcRow>> {
        //酷狗歌词
        return HttpClient.getKuGouApiservice().getKuGouSearch(1, "yes", "pc", mSearchKey, mSong.duration, "")
                .flatMap { body ->
                    val searchResponse = Gson().fromJson(body.string(), KSearchResponse::class.java)
                    HttpClient.getKuGouApiservice().getKuGouLyric(1, "pc", "lrc", "utf8",
                            searchResponse.candidates[0].id,
                            searchResponse.candidates[0].accesskey)
                            .map { lrcBody ->
                                val lrcResponse = Gson().fromJson(lrcBody.string(), KLrcResponse::class.java)
                                LogUtil.d(TAG, "KugouLyric")
                                mLrcParser.getLrcRows(getBufferReader(Base64.decode(lrcResponse.content, Base64.DEFAULT)), true, mCacheKey, mSearchKey)
                            }
                }
    }

    /**
     * 在线歌词
     *
     * @param type
     * @return
     */
    private fun getNetworkObservable(type: Int): Observable<List<LrcRow>> {
        var newType = type

        if (TextUtils.isEmpty(mSearchKey)) {
            return Observable.error(Throwable("no available key"))
        }
        if (newType == SPUtil.LYRIC_KEY.LYRIC_DEFAULT)
            newType = SPUtil.LYRIC_KEY.LYRIC_NETEASE
        if (newType == SPUtil.LYRIC_KEY.LYRIC_KUGOU) {
            //酷狗歌词
            return getKuGouObservable()
        } else {
            //网易歌词
            return getNeteaseObservable()
        }
    }

    private fun isTranslateCanUse(translate: String): Boolean {
        return !TextUtils.isEmpty(translate) && !translate.startsWith("词") && !translate.startsWith("曲")
    }

    /**
     * 获得搜索歌词的关键字
     *
     * @param song
     * @return
     */
    private fun getLyricSearchKey(song: Song?): String {
        if (song == null)
            return ""
        val isTitleAvailable = !ImageUriUtil.isSongNameUnknownOrEmpty(song.title)
        val isAlbumAvailable = !ImageUriUtil.isAlbumNameUnknownOrEmpty(song.album)
        val isArtistAvailable = !ImageUriUtil.isArtistNameUnknownOrEmpty(song.artist)

        //歌曲名合法
        return if (isTitleAvailable) {
            when {
                isArtistAvailable -> song.artist + "-" + song.title //艺术家合法
                isAlbumAvailable -> //专辑名合法
                    song.album + "-" + song.title
                else -> song.title
            }
        } else ""
    }

    @Throws(FileNotFoundException::class, UnsupportedEncodingException::class)
    private fun getBufferReader(path: String): BufferedReader {
        return BufferedReader(InputStreamReader(FileInputStream(path), LyricUtil.getCharset(path)))
    }

    private fun getBufferReader(bytes: ByteArray): BufferedReader {
        return BufferedReader(InputStreamReader(ByteArrayInputStream(bytes), UTF_8))
    }

    companion object {
        private const val TAG = "SearchLrc"

        @JvmStatic
        lateinit var LOCAL_LYRIC_PATH: String

        private val UTF_8 = Charset.forName("UTF-8")
    }
}
